File: backend/hikeway/src/main/java/ua/nure/kryvko/hikeway/model/PointOfInterest.java
package ua.nure.kryvko.hikeway.model;

import jakarta.persistence.*;
import lombok.Data;
import org.locationtech.jts.geom.Point;

import java.time.Instant;

@Entity
@Table(indexes = {
        @Index(name = "idx_poi_owner", columnList = "ownerId"),
        @Index(name = "idx_poi_deleted", columnList = "deleted")
})
@Data
public class PointOfInterest {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String ownerId;

    @Column(nullable = false)
    private String ownerDisplayName;

    @Column(nullable = false, length = 120)
    private String name;

    @Column(nullable = false, length = 4000)
    private String description;

    @Column(nullable = false, columnDefinition = "geometry(Point, 4326)")
    private Point location;

    @Column(nullable = false)
    private Instant createdAt;

    @Column(nullable = false)
    private Instant updatedAt;

    @Column(nullable = false)
    private boolean deleted;
}

File: mobile/app/src/main/java/ua/nure/kryvko/hikeway/core/model/PointOfInterest.kt
package ua.nure.kryvko.hikeway.core.model

import java.time.Instant

data class PointOfInterest(
    val id: Long,
    val name: String,
    val description: String,
    val location: GeoPoint,
    val ownerId: String? = null,
    val ownerDisplayName: String? = null,
    val ownedByCurrentUser: Boolean = false,
    val photos: List<PoiPhoto> = emptyList(),
    val photoResIds: List<Int> = emptyList(),
    val averageRating: Double = 0.0,
    val ratingCount: Long = 0L,
    val userRating: Int? = null,
    val distanceMeters: Double? = null,
    val comments: List<PoiComment> = emptyList(),
    val createdAt: Instant? = null,
    val updatedAt: Instant? = null,
)

data class PoiPhoto(
    val id: Long,
    val contributorId: String,
    val contributorDisplayName: String,
    val ownedByCurrentUser: Boolean,
    val url: String,
    val contentType: String,
    val sizeBytes: Long,
    val caption: String?,
    val createdAt: Instant? = null,
)

data class PoiComment(
    val id: Long,
    val authorId: String,
    val authorDisplayName: String,
    val ownedByCurrentUser: Boolean,
    val text: String,
    val createdAt: Instant? = null,
    val updatedAt: Instant? = null,
)

data class PoiRating(
    val averageRating: Double,
    val ratingCount: Long,
    val userRating: Int?,
)

data class PoiPhotoUpload(
    val bytes: ByteArray,
    val contentType: String,
    val caption: String?,
) {
    val sizeBytes: Long = bytes.size.toLong()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as PoiPhotoUpload
        return bytes.contentEquals(other.bytes) &&
            contentType == other.contentType &&
            caption == other.caption
    }

    override fun hashCode(): Int {
        var result = bytes.contentHashCode()
        result = 31 * result + contentType.hashCode()
        result = 31 * result + (caption?.hashCode() ?: 0)
        return result
    }
}
